File: tobas_setup_assistant/src/tobas_setup_assistant/parameter_getters/line_edit.py
from typing import override
from typing import Optional
from PyQt5.QtCore import pyqtSignal, pyqtSlot
from PyQt5.QtWidgets import QLineEdit

from .base import ParamGetterWidget


class ParamGetterWidget_LineEdit(ParamGetterWidget[str]):
    text_changed = pyqtSignal(str)

    def __init__(self, param_name: str, description_text: Optional[str] = None, default: str = "") -> None:
        super().__init__(param_name, description_text)

        self._line = QLineEdit(default)
        self._rows.addWidget(self._line)

        self._line.textChanged.connect(self._on_text_changed)

    @override
    def get(self) -> str:
        return self._line.text()

    @override
    def set(self, src: str) -> None:
        self._line.setText(src)

    @pyqtSlot(str)
    def _on_text_changed(self, text: str) -> None:
        self.text_changed.emit(text)

File: tobas_gui/tobas_setup_assistant/src/param_getters/line_edit.cpp
#include "tobas_setup_assistant/param_getters/line_edit.hpp"

namespace gui
{
namespace setup_assistant
{
ParamGetterWidget_LineEdit::ParamGetterWidget_LineEdit(
  const QString& param_name,
  const QString& description_text,
  const QString& _default)
  : super(param_name, description_text)
{
  line_ = new QLineEdit(_default);
  rows_->addWidget(line_);

  connect(line_, &QLineEdit::textChanged, this, &self::onTextChanged);
}

QString ParamGetterWidget_LineEdit::get() const
{
  return line_->text();
}

bool ParamGetterWidget_LineEdit::set(const QString& src)
{
  line_->setText(src);
  return true;
}

void ParamGetterWidget_LineEdit::onTextChanged(const QString& text)
{
  Q_EMIT textChanged(text);
}
}  // namespace setup_assistant
}  // namespace gui
